File: December 2022/Burst Balloons/C++/Burst Balloons.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
public:
	int maxCoins(int N, vector<int> &arr)
	{
		arr.insert(arr.begin(), 1);
		arr.push_back(1);
		vector<vector<int>> dp(arr.size(), vector<int>(arr.size(), 0));
		for (int len = 1; len <= N; ++len)
			for (int left = 1; left <= N - len + 1; ++left)
			{
				int right = left + len - 1;
				for (int k = left; k <= right; ++k)
					dp[left][right] = max(dp[left][right], arr[left - 1] * arr[k] * arr[right + 1] + dp[left][k - 1] + dp[k + 1][right]);
			}
		return dp[1][N];
	}
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin>>N;
        
        vector<int> arr(N);
        for(int i=0; i<N; i++)
            cin>>arr[i];

        Solution obj;
        cout << obj.maxCoins(N, arr) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: December 2022/Burst Balloons/Python/Burst Balloons.py
from typing import List

class Solution:
    def maxCoins(self, N : int, arr : List[int]) -> int:
        arr=[1]+arr
        arr.append(1)
        dp=[[0]*(len(arr)+1) for _ in range(len(arr))]
        for le in range(1,N+1):
            for left in range(1,N-le+2):
                right=left+le-1
                for k in range(left,right+1):
                    dp[left][right]=max(dp[left][right],arr[left-1]*arr[k]*arr[right+1]+dp[left][k-1]+dp[k+1][right])
        return dp[1][N]

#{ 
 # Driver Code Starts

class IntArray:
    def __init__(self) -> None:
        pass
    def Input(self,n):
        arr=[int(i) for i in input().strip().split()]#array input
        return arr
    def Print(self,arr):
        for i in arr:
            print(i,end=" ")
        print()


if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        N = int(input())
        
        
        arr=IntArray().Input(N)
        
        obj = Solution()
        res = obj.maxCoins(N, arr)
        
        print(res)
        

# } Driver Code Ends